src/descriptor/mod.rs

```diff
@@ -68,6 +73,126 @@ pub enum Descriptor<Pk: MiniscriptKey> {
     ShWsh(Miniscript<Pk, Segwitv0>),
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub enum DescriptorPublicKey {
+    PubKey(bitcoin::PublicKey),
+    XPub(DescriptorXPub),
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct DescriptorXPub {
+    pub origin: Option<(bip32::Fingerprint, bip32::DerivationPath)>,
+    pub xpub: bip32::ExtendedPubKey,
+    pub derivation_path: bip32::DerivationPath,
+    pub is_wildcard: bool,
```

Comment:
According to https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md we can end with `/*` or `/*'` ... so I think this should be an enum rather than a bool, with values `Fixed`, `Wildcard`, `WildcardHardened` say

Reply:
As long as we don't support private keys in descriptors that doesn't make a lot of sense imo and would only add error cases.

But the stronger argument against this might be the current API design: it allows the asterisk to be replaced by arbitrary `DerivationPath`s, which may contain hardened and unhardened `ChildNumber`s. This design is intentional, it allows one descriptor to describe both the external and the internal wallet by deriving with `m/0/*` and `m/1/*` respectively. As long as we want to support derivation paths longer than 1 the `Wildcard`/`HardenedWildcard` distinction doesn't make any sense (what about e.g. `m/1'/42`? it's both hardened and unhardened).

Reply:
I think we can for now explicitly fail because of hardened wildcard during parsing. What about:
```patch
diff --git a/src/descriptor/mod.rs b/src/descriptor/mod.rs
index 4abfcb2..07f0bb9 100644
--- a/src/descriptor/mod.rs
+++ b/src/descriptor/mod.rs
@@ -240,6 +240,10 @@ impl DescriptorPublicKey {
                 if !is_wildcard && p == "*" {
                     is_wildcard = true;
                     None
+                } else if !is_wildcard && p == "*'" {
+                    Some(Err(DescriptorKeyParseError(
+                        "Hardened derivation is currently not supported.",
+                    )))
                 } else if is_wildcard {
                     Some(Err(DescriptorKeyParseError(
                         "'*' may only appear as last element in a derivation path.",
@@ -1457,6 +1461,15 @@ mod tests {
             ))
         );
 
+        // And ones which claim to be able to derive hardened child
+        let desc = "[78412e3a/44'/0'/0']xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL/1/42/*'";
+        assert_eq!(
+            DescriptorPublicKey::from_str(desc),
+            Err(DescriptorKeyParseError(
+                "Hardened derivation is currently not supported."
+            ))
+        );
+
         // And ones with misplaced wildcard
         let desc = "[78412e3a/44'/0'/0']xpub6ERApfZwUNrhLCkDtcHTcxd75RbzS1ed54G1LkBUHQVHQKqhMkhgbmJbZRkrgZw4koxb5JaHWkY4ALHY2grBGRjaDMzQLcgJvLJuZZvRcEL/1/*/44";
         assert_eq!(
```

Reply:
Yeah, maybe we should also describe the deviation from core's behavior in the docs. It feels intuitive for me, but might not for others.

Reply:
Definitely :-)
Would be great to have it in the release that Sanket and Andrew have been talking about on IRC as well